flow/BlobCipher.h

```diff
@@ -0,0 +1,244 @@
+	uint64_t getCreationTime() const { return creationTime; }
+	BlobCipherDomainId getDomainId() const { return encryptDomainId; }
+	BlobCipherRandomSalt getSalt() const { return randomSalt; }
+	BlobCipherBaseKeyId getBaseCipherId() const { return baseCipherId; }
+	int getBaseCipherLen() const { return baseCipherLen; }
+	uint8_t* rawCipher() const { return cipher.get(); }
+	uint8_t* rawBaseCipher() const { return baseCipher.get(); }
+	void reset();
+};
+
+// This interface allows FDB processes participating in encyrption to store and index recently used encyption cipher
+// keys. FDB encryption has three dimensions:
+// 1. Mapping on cipher encyrption keys per "encryption domains"
+// 2. Per encryption domain, the cipher keys are index using "baseCipherKeyId"
+// 3. Within baseCipherKey ids indexed encryption keys, cipher keys are indexed based on the "randomSalt" used to apply
+// HMAC-SHA256 derivation.
+//                  { encryptionDomain -> { baseCipherId -> { randomSalt, cipherKey } } }
+//
+// Supported cache lookups schemes:
+// 1. Lookup cipher based on { encyrptionDomainId, baseCipherKeyId } tuple.
+// 2. Lookup cipher based on BlobCipherEncryptionHeader
+//
+// Client is responsible to handle cache-miss usecase, the corrective operation might vary based on the
+// calling process, for instance: EncryptKeyServer cache-miss shall invoke RPC to external Encryption Key Manager to
+// fetch the required encryption key, however, CPs/SSs cache-miss would result in RPC to EncryptKeyServer to refresh the
+// desired encryption key.
+
+using BlobCipherKeySaltCache = std::unordered_map<BlobCipherRandomSalt, Reference<BlobCipherKey>>;
+
+class BlobCipherKeyItem : public ReferenceCounted<BlobCipherKeyItem>, NonCopyable {
+	Reference<BlobCipherKey> latest;
+	BlobCipherKeySaltCache keyCache;
+
+public:
+	BlobCipherKeyItem(Reference<BlobCipherKey>& cipher) { updateLatest(cipher); }
+
+	Reference<BlobCipherKey> getLatest() { return latest; }
+
+	void updateLatest(Reference<BlobCipherKey> cipher) {
+		latest = cipher;
+		keyCache.emplace(cipher.getPtr()->getSalt(), cipher);
+	}
+	Reference<BlobCipherKey> findCipher(const BlobCipherRandomSalt& salt) {
+		auto itr = keyCache.find(salt);
+		if (itr == keyCache.end()) {
+			return Reference<BlobCipherKey>();
+		}
+		return itr->second;
+	}
+	void reset() {
+		for (auto& keyItr : keyCache) {
+			keyItr.second.getPtr()->reset();
+		}
+	}
+	std::vector<Reference<BlobCipherKey>> getAllCiphers() {
+		std::vector<Reference<BlobCipherKey>> ciphers;
+		for (auto itr : keyCache) {
+			ciphers.emplace_back(itr.second);
+		}
+		return ciphers;
+	}
+};
+
+using BlobCipherKeyIdCacheMap = std::unordered_map<BlobCipherBaseKeyId, Reference<BlobCipherKeyItem>>;
+using BlobCipherDomainCacheMap = std::unordered_map<BlobCipherDomainId, BlobCipherKeyIdCacheMap>;
+
+class BlobCipherKeyCache : NonCopyable {
+	BlobCipherDomainCacheMap domainCacheMap;
+	static constexpr uint64_t CIPHER_KEY_CACHE_TTL_NS = 10 * 60 * 1000 * 1000 * 1;
```

Comment:
I'm confused about the unit here. The TTL logic in BlobCipherKeyCache is using fdb's now(), which has the unit of seconds, and this says it's in NS? If this is supposed to be 10 minutes, and are using fdb's now(), I think you want this to be a double of 600.0?

Reply:
Ah! This seems to be fallout from my change, I remember updating this to use `chrono`, but given the above comment will convert usage to leverage timer_monotonic. 
Thanks